hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/SerialNumberManager.java

```diff
@@ -115,6 +119,36 @@ public static StringTable getStringTable() {
     return map;
   }
 
+  // returns serial snapshot of current values for a save.
+  public static StringTable getSerialStringTable() {
+    // approximate size for capacity.
+    int size = 0;
+    for (final SerialNumberManager snm : values) {
+      size += snm.size();
+    }
+
+    int tableMaskBits = getMaskBits();
+    StringTable map = new StringTable(size, 0);
+    Set<String> entrySet = new HashSet<String>();
+    AtomicInteger index = new AtomicInteger();
+
+    Map <Integer, String> entryMap = new TreeMap<Integer, String>();
+    for (final SerialNumberManager snm : values) {
+      final int mask = snm.getMask(tableMaskBits);
+      for (Entry<Integer, String> entry : snm.entrySet()) {
+        entryMap.put(entry.getKey() | mask, entry.getValue());
+      }
+    }
+
+    entryMap.forEach((key, value) -> {
+      if(entrySet.add(value)){
+        map.put(index.incrementAndGet(), value);
```

Comment:
Why not execute the incrementAndGet method when persistently collecting information?

Reply:
incrementAndGet method is executed when the fsimage file is persisted to ensure that the index is continuous.